apiserver/facades/client/modelmanager/modelmanager.go

```diff
@@ -806,55 +806,67 @@ func (m *ModelManagerAPI) ListModelSummaries(ctx context.Context, req params.Mod
 		return result, errors.Trace(err)
 	}
 
-	modelInfos, err := m.state.ModelSummariesForUser(userTag, req.All && m.isAdmin)
+	var modelInfos []model.ModelSummary
+	if req.All {
+		if !m.isAdmin {
+			return result, fmt.Errorf("%w: cannot list all models as non-admin user", apiservererrors.ErrPerm)
+		}
+		modelInfos, err = m.modelService.ListAllModelSummaries(ctx)
+	} else {
+		modelInfos, err = m.modelService.ListModelSummariesForUser(ctx, userTag.Id())
+	}
 	if err != nil {
 		return result, errors.Trace(err)
 	}
+
 	err = m.fillInStatusBasedOnCloudCredentialValidity(ctx, modelInfos)
 	if err != nil {
-		return result, errors.Trace(err)
+		return result, apiservererrors.ServerError(
+			errors.Annotatef(err, "listing model summaries: filling in status for missing cloud credential"),
+		)
 	}
 
 	for _, mi := range modelInfos {
-		lastConnection, err := m.accessService.LastModelLogin(ctx, userTag.Id(), coremodel.UUID(mi.UUID))
-		if errors.Is(err, accesserrors.UserNeverAccessedModel) {
-			mi.UserLastConnection = nil
-		} else if errors.Is(err, modelerrors.NotFound) {
-			// TODO (aflynn): Once models are fully in domain, replace the line
-			// below with a `continue`. When models are still in state, this
-			// case is triggered because the model cannot be found in the domain
-			// db. Generally, it should only be triggered if the model has been
-			// removed since we got the UUID.
-			mi.UserLastConnection = nil
-		} else if err != nil {
-			return result, errors.Annotatef(err, "getting model last login time for user %q on model %q", userTag.Name(), mi.Name)
+		summary, err := m.makeModelSummary(mi)
+		if err != nil {
+			result.Results = append(result.Results, params.ModelSummaryResult{Error: apiservererrors.ServerError(err)})
 		} else {
-			mi.UserLastConnection = &lastConnection
+			result.Results = append(result.Results, params.ModelSummaryResult{Result: summary})
 		}
-		summary := m.makeModelSummary(mi)
-		result.Results = append(result.Results, params.ModelSummaryResult{Result: summary})
 	}
 	return result, nil
 
 }
 
-func (m *ModelManagerAPI) makeModelSummary(mi state.ModelSummary) *params.ModelSummary {
+func (m *ModelManagerAPI) makeModelSummary(mi model.ModelSummary) (*params.ModelSummary, error) {
+	credTag, err := mi.CloudCredentialKey.Tag()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if !names.IsValidCloud(mi.CloudName) {
+		return nil, fmt.Errorf("invalid cloud name %q", mi.CloudName)
+	}
+	cloudTag := names.NewCloudTag(mi.CloudName)
+	if !names.IsValidUser(mi.OwnerName) {
```

Comment:
Same comments as above. Names shouldn't be prescribing business logic like this when we have already accepted what a valid name is.

Reply:
Same reasoning as above